# AOC Day 2!

### Hannah Valenty

Checking to be sure all levels (row) in the text file are increasing or decreasing gradually.

* Adjacent levels can differ by at least one and at most three (changing by 1-3).
* Must be entirely increasing or decreasing.
* Also cannot have numbers that neither increase nor decrease.

In [1]:
import numpy as np
import pandas as pd

In [10]:
# open and read txt file
with open('../data/day02_input.txt', 'r') as f:
    file_lines = f.readlines()
# remove newline character
file_lines = [string.replace('\n', '') for string in file_lines]
# transform read data into pandas dataframe
keymap =  pd.DataFrame([string.split(' ') for string in file_lines])


In [11]:
keymap

0   1   2   3   4     5     6     7
0     1   4   5   8  11    12     9  None
1     7   8   9  10  12    15    17    17
2    17  20  23  25  27    31  None  None
3    55  57  58  61  63    64    70  None
4    39  42  45  43  44  None  None  None
..   ..  ..  ..  ..  ..   ...   ...   ...
995  91  89  87  85  82    81    78  None
996  77  79  81  82  85    86  None  None
997  59  61  62  65  67    68  None  None
998  94  93  92  91  90    87    85    83
999  50  52  53  54  56    57    58    61

[1000 rows x 8 columns]

### Part 1: Counting the number of rows which meet the conditions above.

In [ ]:
# tried to avoid the nested for loop but just went with it!

# counter for final number
total = 0

# loop through each row of dataframe
for row in range(len(keymap)):

    direction = []
    change = []

    # loop to check
    for element in range(len(keymap.loc[row])-1):
        # assign values to compare
        first = keymap.loc[row][element]
        element2 = 1 + element
        second = keymap.loc[row][element2]
        if second == None:
            break
        # subtract elements for comparison
        diff = int(first) - int(second)
        # set if increasing of decreasing 
        if diff < 0:
            direction.append('increasing')
        elif diff > 0:
            direction.append('decreasing')
        elif diff == 0:
            direction.append('no change')
        # set for amount of change
        change.append(abs(diff))

    # check for change in direction and amount
    dir_result = len(set(direction)) == 1
    ch_result = all([1 <= value <= 3 for value in change])

    if dir_result == True & ch_result == True:
        total+=1



In [81]:
total

549

**Final Answer:** 549 (549 lines meet the conditions)

### Part 2: Count lines for updated conditions

* One aspect of the original conditions may be violated to still count the row
* Allow the new row to not violate the conditions by removing one entry